server/src/services/InitGoogle.ts

```diff
@@ -92,41 +100,46 @@ async function onTokens(tokens: Credentials) {
       update,
       create: { id: TOKENS_ID, ...update },
     });
-  } else {
+  } else if (existingGoogleTokens) {
     const update = { access_token, expiry_date, is_valid: true };
-    // TODO: Handle the case where the refresh token is not sent *and* the
-    // record doesn't exist. If this happens, we need to redirect them to a
-    // new auth url, but with prompt: 'consent', so that Google will provide
-    // a new refresh token.
     await prisma.google_tokens.update({
       where: { id: TOKENS_ID },
       data: { ...update },
     });
+  } else {
+    // This should not happen, but if it did, presumably something went
+    // wrong during the Google authentication flow. All we can do is ask the
+    // user to try again.
+    throw new Error('Missing refresh_token');
   }
 }
 
 export async function requestTokens(code: string) {
-  const oauth2Client = createOAuth2Client().on('tokens', onTokens);
-
+  const oauth2Client = createOAuth2Client();
   try {
-    await oauth2Client.getToken(code);
-  } catch {
-    throw new Error('Failed to get tokens');
+    const tokens = (await oauth2Client.getToken(code)).tokens;
+    await onTokens(tokens, { requireRefreshToken: true });
+  } catch (e) {
+    console.error('Failed to get tokens');
+    throw redactSecrets(e);
   }
 }
 
-export function getGoogleAuthUrl(state: string) {
+export function getGoogleAuthUrl(state: string, prompt = false) {
   const oauth2Client = createOAuth2Client();
 
   return oauth2Client.generateAuthUrl({
     access_type: 'offline',
     scope: scopes.join(' '),
     state,
+    ...(prompt && { prompt: 'consent' }),
   });
 }
 
 async function createCredentialedClient() {
-  const oauth2Client = createOAuth2Client().on('tokens', onTokens);
+  const oauth2Client = createOAuth2Client().on('tokens', (tokens) =>
+    onTokens(tokens),
+  );
```

Comment:
```suggestion
  const oauth2Client = createOAuth2Client().on('tokens', onTokens);
```
TS is clever enough to recognise that the second argument is optional, so there's no need to wrap the function call.

Reply:
Thanks, sometimes I'm getting in a create-functions-everywhere mode.